_posts/2024-06-14-remote-code-execution-engine.md

```diff
@@ -0,0 +1,1426 @@
+
+### 4.1 transformation.conf
+
+The project also requires a special configuration for capturing properties for domain objects such as `worker`, `load-balancer` and so on.
+
+```hocon
+include "application"
+
+transformation {
+  workers-per-node = 32
+  load-balancer = 3
+}
+```
+Here, it simply means that each node will have 32 worker actors and master node will have 3 load balancer actors.
+In real world, choosing those numbers would depend on multiple variables that must be collected and analyzed in production.
+In my opinion, those numbers are optimized based on empirical evidence rather than theoretical results.
+
+### 4.2 Serialization
+
+Since we're building a cluster we need to keep in mind that actor messages must be serialized over wire. It means that messages could be sent from JVM to JVM which requires enabling the serialization protocol. 
+
+For that we need a simple trait definition under `serialization` package, as we've already defined in configuration.
+
+```scala
+package serialization
+
+// just a marker trait to tell pekko to serialize messages into CBOR using Jackson for sending over the network
+trait CborSerializable
+```
+
+## 5. Actors
+
+In this section we will write the code for actors that will do the real job - code execution.
+
+Those actors will be:
+- `Worker`: actor that initiates code execution
+- `FileHandler`: child actor of `Worker` which prepares the temporary file before executing it
+- `CodeExecutor`: child actor of `FileHandler` which starts the process and collects output
+
+Now, time to dig into the details.
+
+### 5.1 Worker
+
+Let's quickly evaluate what we would expect from `Worker`.
+
+`Worker` actor will live on `worker` node and will be reacting to `StartExecution` message. 
+
+Also, during runtime, it must be aware of which languages are supported, so that it quickly replies something like `ExecutionFailed` response with the message `"unsupported language"`.
+
+For successful case, we could have `ExecutionSucceeded` message that will be forwarded back to the original sender.
+
+How about making `Worker` actor work within a cluster? We said earlier that the `Worker` actor will be expecting messages from `master` node, here enters remoting and other good stuff. 
+
+Somehow, `Worker` actor must be discovered within `pekko` cluster so that messages can be routed to it, for that, we'll be using `ServiceKey` and later, `Router`-s which are special actors.
+
+So, keeping all this in mind the code could look like this:
+
+```scala
+package workers
+
+import org.apache.pekko.actor.typed.receptionist.ServiceKey
+import workers.children.FileHandler.In.PrepareFile
+import org.apache.pekko.actor.typed.{ActorRef, Behavior}
+import org.apache.pekko.actor.typed.scaladsl.Behaviors
+import serialization.CborSerializable
+import workers.children.FileHandler
+
+import scala.util.*
+
+object Worker:
```

Comment:
Avoid the braceless syntax on blog posts - it causes issues when people copy the code locally.

I know it's annoying to change it, but I've had several issues myself when replicating articles in my own dev environment, which is why I have replaced everything with no-indent.